snow/engine/snowman/event/queue_test.go

```diff
@@ -0,0 +1,294 @@
+	execute func(context.Context) error
+	cancel  func(context.Context) error
+}
+
+func (j *testJob) Execute(ctx context.Context) error {
+	return j.execute(ctx)
+}
+
+func (j *testJob) Cancel(ctx context.Context) error {
+	return j.cancel(ctx)
+}
+
+func newQueueWithJob[T comparable](t *testing.T, job Job, shouldCancel bool, dependencies ...T) *Queue[T] {
+	q := NewQueue[T]()
+	require.NoError(t, q.Register(context.Background(), job, dependencies...))
+	if shouldCancel {
+		for _, jobs := range q.jobs {
+			for _, j := range jobs {
+				j.shouldCancel = true
+			}
+		}
+	}
+	return q
+}
+
+func TestQueue_Register(t *testing.T) {
+	var calledExecute bool
+	userJob := &testJob{
+		execute: func(context.Context) error {
+			if calledExecute {
+				return errDuplicateInvocation
+			}
+			calledExecute = true
+			return nil
+		},
+		cancel: func(context.Context) error {
+			return errUnexpectedInvocation
+		},
+	}
+
+	tests := []struct {
+		name          string
+		queue         *Queue[int]
+		dependencies  []int
+		shouldExecute bool
+		expectedLen   int
+		expectedQueue *Queue[int]
+	}{
+		{
+			name:          "no dependencies",
+			queue:         NewQueue[int](),
+			dependencies:  nil,
+			shouldExecute: true,
+			expectedLen:   0,
+			expectedQueue: NewQueue[int](),
+		},
+		{
+			name:          "one dependency",
+			queue:         NewQueue[int](),
+			dependencies:  []int{1},
+			shouldExecute: false,
+			expectedLen:   1,
+			expectedQueue: &Queue[int]{
+				jobs: map[int][]*job[int]{
+					1: {
+						{
+							dependencies: set.Of(1),
+							job:          userJob,
+						},
+					},
+				},
+			},
+		},
+		{
+			name:          "two dependencies",
+			queue:         NewQueue[int](),
+			dependencies:  []int{1, 2},
+			shouldExecute: false,
+			expectedLen:   2,
+			expectedQueue: &Queue[int]{
+				jobs: map[int][]*job[int]{
+					1: {
+						{
+							dependencies: set.Of(1, 2),
+							job:          userJob,
+						},
+					},
+					2: {
+						{
+							dependencies: set.Of(1, 2),
+							job:          userJob,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "additional dependency",
+			queue: &Queue[int]{
+				jobs: map[int][]*job[int]{
+					1: {
+						{
+							dependencies: set.Of(1),
+							job:          userJob,
+						},
+					},
+				},
+			},
+			dependencies:  []int{1},
+			shouldExecute: false,
+			expectedLen:   1,
+			expectedQueue: &Queue[int]{
+				jobs: map[int][]*job[int]{
+					1: {
+						{
+							dependencies: set.Of(1),
+							job:          userJob,
+						},
+						{
+							dependencies: set.Of(1),
+							job:          userJob,
+						},
+					},
+				},
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			require := require.New(t)
+
+			// Reset the variable between tests
+			calledExecute = false
+
+			require.NoError(test.queue.Register(context.Background(), userJob, test.dependencies...))
+			require.Equal(test.expectedLen, test.queue.NumDependencies())
```

Comment:
`require.X()` will terminate early and create a whack-a-mole testing experience, so `assert.X()` is more appropriate. It makes sense to still use `require` for `NoError()` though, as that is guaranteed to invalidate all future tests.

This applies to all tests.

[See style guide](https://google.github.io/styleguide/go/decisions#keep-going)

Reply:
Currently we actually explicitly ban the usage of `assert` in favor of `require`: https://github.com/ava-labs/avalanchego/blob/2e72c7c29c498e7ffa5bb31ff18495e14e2cdfb7/.golangci.yml#L93-L94

In the past we had the issue (especially in some of our more, _ahem_, "legacy" tests) where assertions were used and caused annoying to debug errors (because the original error was lost in the weeds).

I think it probably makes sense to pull this conversation out of this PR and then we can discuss:
1. Unbanning `assert`.
2. Doing a pass to use `assert` where it makes sense.